tests/VimPlugin/fileutils.vader

```diff
@@ -0,0 +1,46 @@
+Before:
+  Save g:fixtureDir
+  Save g:project1Dir
+  Save g:subdir
+  Save g:marker
+  Save g:markerFilePath
+  let g:fixtureDir = tempname()
+  let g:project1Dir = g:fixtureDir . '/project1'
+  let g:subdir = g:project1Dir . '/foo/bar'
+  let g:marker = 'composer.json'
+  let g:markerFilePath = g:project1Dir . '/' . g:marker
+
+  call mkdir(g:subdir, 'p')
+  execute "write " . g:markerFilePath
+
+Execute (Test search upward for the nearest directory containing given file as a g:marker):
+  AssertEqual g:project1Dir, phpactor#fileutils#searchUpwardBySingleMarker(g:project1Dir, g:marker)
+  AssertEqual g:project1Dir, phpactor#fileutils#searchUpwardBySingleMarker(g:markerFilePath, g:marker)
+  AssertEqual g:project1Dir, phpactor#fileutils#searchUpwardBySingleMarker(g:subdir, g:marker)
+  AssertEqual v:null, phpactor#fileutils#searchUpwardBySingleMarker(g:fixtureDir, g:marker)
+
+  execute "write " . g:project1Dir . '/foo/' .g:marker
+  AssertEqual g:project1Dir . '/foo', phpactor#fileutils#searchUpwardBySingleMarker(g:subdir, g:marker)
+
+Execute (Test phpactor#fileutils#isSubdir()):
+  Assert phpactor#fileutils#isSubdir('/usr/bin', '/usr')
+  Assert phpactor#fileutils#isSubdir('/usr/', '/usr')
+  Assert ! phpactor#fileutils#isSubdir('/', '/usr')
+
+Execute (Test searching upward given root markers and forbidden dirs):
+  let rootMarkers = ['bar', g:marker]
+  let forbiddenDirs = []
+
+  AssertEqual g:project1Dir, phpactor#fileutils#searchDirectoryUpwardByRootMarkers(g:subdir, rootMarkers, forbiddenDirs)
+
+  let forbiddenDirs = [g:project1Dir.'/foo']
+  AssertEqual v:null, phpactor#fileutils#searchDirectoryUpwardByRootMarkers(g:subdir, rootMarkers, forbiddenDirs)
```

Comment:
The purpose of excluding some directory from be a project root is an attempt to avoid indexing large directory that is supposed to be not a project root. `/` and '$HOME` are usual examples. If searching upward stops at one of excluded directories then there is no reasonable value to suggest and another resolver should be used. In non-interactive variant (`g:phpactorAllowInteractiveProjectResolution == v:false`) the next resolver will be used (`initialCwd` is always present independently of user settings).

It is not intended to disable any marker from be stop in moving upward so I agree with your scenario except the last assertion. 

However, if a fallback directory (initialCwd) is one of excluded (e.g. we start from `~/`) and will be used, then we cannot avoid it.

Reply:
The case added by you is a good example to show that searching upward to any root marker is not always a good resolver. Assuming additionally that we start (and still have CWD) in home directory or one of excluded, we also have a good example to show that `initialCwd` is not god resolver for this case. Thus I added `manual` method and option to enable interactive choice from all methods.

Reply:
OK so it's not meant to say "Ignore the vendor as project root and search upwards", it meant to say "Stop right now there is no interest to look anymore".

At first I thought it was an option to avoid taking the vendor directory into account, hence my misunderstanding.